### PR TITLE
feat: add resources conditionally

### LINKED ISSUE #313: Add resources conditionally
this issue is converted from this [RFC PR](https://github.com/odpf/guardian/pull/156) 

# Summary

Resource in Guardian represents the actual resource in the provider e.g. for BigQuery provider, a resource represents a
dataset or a table. Guardian collects resources from the provider automatically as soon as it registered. While in
parallel, Guardian also has a job for continuously syncing resources.

In current design it fetch all the database/table/resources in configure provider, there is no way to filter out on
resources.

# Technical Design

Add a scope of condition in provider configuration to filter out on resources.

#### Using expression

Expression can be used to check whether Guardian should add the resources or not.

```yaml
resources:
  filter: $resource.urn endsWith 'keyset'
  type: tables
  policy:
    id: keyset_table_policy
```

#### Migration

We can edit or add new providers based on new configuration. For already added resources whom we want to make conditional need to soft-delete it from Guardian db.

### PR DESCRIPTION
**Pull request by singhvikash11**
Closes  #313

### CODE REVIEW

File: domain/provider.go [hunk 1]
```diff
@@ -40,6 +40,7 @@ type PolicyConfig struct {
 // ResourceConfig is the configuration for a resource type within a provider
 type ResourceConfig struct {
 	Type   string        `json:"type" yaml:"type" validate:"required"`
+	Filter string        `json:"filter" yaml:"type"`
```
Comment (bsushmith) [inline domain/provider.go:43]: yaml tag is incorrect.

File: docs/docs/providers/bigquery.md [hunk 2]
```diff
@@ -71,6 +71,7 @@ resources:
         permissions:
           - WRITER
   - type: table
+    filter: $resource.name endsWith transtaction
```
Comment (bsushmith) [inline docs/docs/providers/bigquery.md:74]: Can you fix the typo in here - `transaction`

File: core/provider/service_test.go [hunk 3]
```diff
@@ -312,6 +312,43 @@ func (s *ServiceTestSuite) TestFetchResources() {
...
 	})
+
+	s.Run("should upsert filter resources on success", func() {
+		providersWithResourceFilter := []*domain.Provider{
+			{
+				ID:   "1",
+				Type: mockProviderType,
+				URN:  mockProvider,
+				Config: &domain.ProviderConfig{Resources: []*domain.ResourceConfig{
+					{Type: "dataset", Filter: "$resource.urn == 'resource2'"},
```
Comment (bsushmith) [inline core/provider/service_test.go:323]: Can you add one more test case with `$resource.name endsWith transaction`  as the filter criteria

File: core/provider/service.go [hunk 4]
```diff
@@ -564,3 +595,20 @@ func validateDuration(d string) error {
 	}
 	return nil
 }
+
+func structToMap(item interface{}) (map[string]interface{}, error) {
```
Comment (bsushmith) [inline core/provider/service.go:599]: Can you move this generic function to utils?

File: core/provider/service.go [hunk 5]
```diff
@@ -450,14 +453,42 @@ func (s *Service) getResources(ctx context.Context, p *domain.Provider) ([]*doma
...
+				return nil, fmt.Errorf("parsing resource struct to map: %w", err)
+			}
+
+			v, err := evaluator.Expression(filterExpression).EvaluateWithVars(map[string]interface{}{
+				"resource": resourceMap,
+			})
+			if err != nil {
+				return nil, err
+			}
+			if !reflect.ValueOf(v).IsZero() {
```
Comment (bsushmith) [inline core/provider/service.go:481]: I'm not sure if this also checks if `v` is nil. Can you please check and add a `IsNil` check for `v`?

Comment (singhvikash11) [reply core/provider/service.go:481]: ValueOf check for nil

**Review by bsushmith — APPROVED** [review]

**Review by rahmatrhd — CHANGES_REQUESTED** [review]
let's add a validation on provider create&update for this new `filter` field. We can create a dummy resource and try to evaluate the expression onto that resource to see if the expression if valid and working or not

File: pkg/evaluator/expression.go [hunk 6]
```diff
@@ -23,6 +24,23 @@ func (m ExprParam) Split(s, sep string) []string {
 	return strings.Split(s, sep)
 }
 
+func (e Expression) EvaluateWithStruct(item interface{}) (interface{}, error) {
```
[the remark this answers is not in the document]
Comment (singhvikash11) [reply pkg/evaluator/expression.go:27]: @mabdh @rahmatrhd test is already there for this
https://github.com/odpf/guardian/blob/filter_resource/core/provider/service_test.go#L323

**Comment by singhvikash11** [comment]
> let's add a validation on provider create&update for this new `filter` field. We can create a dummy resource and try to evaluate the expression onto that resource to see if the expression if valid and working or not

https://github.com/odpf/guardian/blob/filter_resource/core/provider/service_test.go#L323
I have verified this manually also

**Comment by singhvikash11** [comment]
@rahmatrhd @mabdh @bsushmith please review changes and approve the pr so we can merge it

**Comment by ravisuhag** [comment]
@rahmatrhd Please see if we can review and close this.

**Review by rahmatrhd — APPROVED** [review]
LGTM

**Pull request merged**